### PR TITLE
Fix cell filtering after clustering switch (SCP-5297)

### PR DESCRIPTION
**Pull request by eweitz**
This makes the cell filtering UX more robust, refactors code for clarity, and adds an automated test.

### Before
Previously, a few bugs occurred when you selected a filter, then went back to the Options panel, then selected another value in the "Clustering" menu.  First, you'd often see an error message `TypeError: Cannot read properties of undefined (reading 'includes')`.  The plot legend would also disappear.  Subsequent cell filtering also showed a permanently stuck "Filtered" legend entry.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/14b19a6a-4cc5-4ba7-b43a-0484d1813ef1

### After
Now, that all works as expected!

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/4b166950-1343-4a3a-a700-680e79d8e448

Another UX refinement: changing the "Color by" annotation no longer clears filters.  Doing so is unnecessary, as cells are the same across annotations.  Preserving filters across color changes allows nimbler exploration.

Behind these UX changes, code quality has also been improved.  Earlier prototype code had generally referred to the "Cell filtering" components as `FacetFilter<Foo>`.  To simplify the cognitive connection between this code and that UI display, the code now refers to the "Cell filtering" components as `CellFiltering<Foo>`, e.g. `CellFilteringPanel`.  This also disambiguates cell filtering code -- which enables faceted search for plotted cells within a study -- to the "Advanced search" code -- which also sometimes uses the term `FacetFilter` in code which enables faceted search for studies across the entire application.

### Test
A new automated test verifies basic initialization of the cell filtering panel React component.  To manually test:
* If on staging, go to [SCP303](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303); if local, go to "Human milk - differential expression" DE pilot
* Enable cell filtering flag if it's not already
* Click "Cell filtering", select a filter, then deselect it, then click the "<-" button to go back to the Options panel
* Select a different clustering in the "Clustering" menu
* Confirm you don't see an error message "TypeError: Cannot read properties of undefined (reading 'includes')`.", and that  the plot legend would appears.

This relates to SCP-5297.

### CODE REVIEW

File: app/javascript/components/explore/CellFilteringPanel.jsx [hunk 1]
```diff
@@ -12,15 +12,15 @@ import { initCellFaceting } from '~/lib/cell-faceting'
...
+export function CellFilteringPanelHeader({
   togglePanel, updateFilteredCells
 }) {
   return (
     <>
       <span> Cell filtering </span>
       <button className="action fa-lg cell-filtering-exit-panel"
         onClick={() => {
-          updateFilteredCells({})
+          updateFilteredCells(null)
```
Comment (eweitz) [inline app/javascript/components/explore/CellFilteringPanel.jsx:23]: Functional change

File: app/javascript/components/explore/ExploreDisplayTabs.jsx [hunk 2]
```diff
@@ -201,6 +201,10 @@ export default function ExploreDisplayTabs({
   /** Update filtered cells to only those that match annotation group value filter selections */
   function updateFilteredCells(selection) {
     if (!cellFaceting) {return}
+    if (!selection) {
+      setFilteredCells(null)
+      return
+    }
```
Comment (eweitz) [inline app/javascript/components/explore/ExploreDisplayTabs.jsx:207]: Functional change

File: app/javascript/components/explore/CellFilteringPanel.jsx [hunk 3]
```diff
@@ -205,7 +205,6 @@ export function FacetFilterPanel({
               setColorByFacet(newColorByAnnotation)
               updateClusterParams({ annotation: newColorByAnnotation })
               currentlyInUseAnnotations.colorBy = newColorByAnnotation
-              updateFilteredCells({})
```
Comment (eweitz) [inline app/javascript/components/explore/CellFilteringPanel.jsx:208]: Functional change

**Review by eweitz — COMMENTED** [review]
To help highlight functional changes among the background of refactoring, I've added "Functional change" comments to specific relevant lines.

**Review by jlchang — COMMENTED** [review]
In doing a functional review, the testing instructions worked as described but a couple of follow up clicks led to another "includes" error. Of course trying to replicate the issue, I could not recall how I caused it. I've attached a regrettably-long video that shows I'm still able to trigger an "includes" error (@51s in first video). 

https://github.com/broadinstitute/single_cell_portal_core/assets/6005195/a1827e6f-b8b7-41ac-aafe-eccefd6068f0


The includes error is followed by showing that I can also trigger a "TypeError: Cannot read properties of undefined (reading 'iconColor')" error (@1m41s in first video), which may be related to separate cell filtering work. The `undefined (reading 'iconColor')` error can be reliably reproduced (see @43s and @1m3s second video) when the selected annotation is a non-default annotation. 

https://github.com/broadinstitute/single_cell_portal_core/assets/6005195/ccc79e43-9133-4cb4-8dab-540dea243023

If these issues should be considered follow-on work in a separate ticket, let me know and I can approve instead of just leaving this comment. I'll try and do more testing to see if I can figure out a quicker path to reproducing the issue.

**Comment by eweitz** [comment]
Great finds, Jean.  

Those issues strike me as worth fixing downstream, as I suspect they're not caused by changes in this PR.  I've opened SCP-5358 for that.  If you'd like to distill steps to reproduce that'd certainly help, but I could also see such diagnostic work being part of that separate ticket.

**Comment by eweitz** [comment]
The [build failure](https://github.com/broadinstitute/single_cell_portal_core/actions/runs/6423887522/job/17443372230?pr=1895) is a false positive, as a Rails test failed but not Rails code changed:

```
Total elapsed time: 34 minutes, 54 seconds

### FAILURES and ERRORS ###
*** Yarn test failures ***

*** Rails test failures ***

  StudyCreationTest#test_create_default_testing_study:
  RuntimeError: Even after 600 seconds, not all files have been parsed.

  SearchFacetPopulatorTest#test_populate_facets_from_alexandria_convention_data:
  NoMethodError: undefined method `dig' for 16279:Integer

  SearchFacetPopulatorTest#test_updates_existing_facets:
  NoMethodError: undefined method `dig' for 16279:Integer
Exiting with code: 1
Error: Process completed with exit code 1.
```

This was likely due to the OLS outage; see [context in Slack](https://broadinstitute.slack.com/archives/G01HY2DJTFY/p1696435959652669).

**Comment by bistline** [comment]
> The [build failure](https://github.com/broadinstitute/single_cell_portal_core/actions/runs/6423887522/job/17443372230?pr=1895) is a false positive, as a Rails test failed but not Rails code changed:
> 
> ```
> Total elapsed time: 34 minutes, 54 seconds
> 
> ### FAILURES and ERRORS ###
> *** Yarn test failures ***
> 
> *** Rails test failures ***
> 
>   StudyCreationTest#test_create_default_testing_study:
>   RuntimeError: Even after 600 seconds, not all files have been parsed.
> 
>   SearchFacetPopulatorTest#test_populate_facets_from_alexandria_convention_data:
>   NoMethodError: undefined method `dig' for 16279:Integer
> 
>   SearchFacetPopulatorTest#test_updates_existing_facets:
>   NoMethodError: undefined method `dig' for 16279:Integer
> Exiting with code: 1
> Error: Process completed with exit code 1.
> ```
> 
> This was likely due to the OLS outage; see [context in Slack](https://broadinstitute.slack.com/archives/G01HY2DJTFY/p1696435959652669).

Yes, that is the cause - I replicated this error locally and the `undefined method 'dig' for 16279:Integer` is due to the response from OLS being in HTML instead of JSON.

**Review by jlchang — APPROVED** [review]
Sounds good. I'll put more details about the UX issues in the follow-on ticket.

**Pull request merged**